src/ServerScriptService/Services/BulletHandlerService/init.lua

```diff
@@ -1,14 +1,228 @@
+        return barrelDistance < barrelMaxCheckDistance
+    end
+
+    local marginOfErrorDistanceIncrement = 0.5
+
+    validationTests["EndPositionCheck"] = function(player: Player, validationData: validationDataType): boolean
+        local bulletData = validationData.BulletData
+        local gunConfig = validationData.GunConfig
+        local barrelDistance = (bulletData.BarrelPosition - bulletData.EndPosition).Magnitude
+        return barrelDistance < (gunConfig.Range+marginOfErrorDistanceIncrement)
+    end
+
+    local marginOfErrorTimeOffset = 0.05
+
+    local function getBulletCount(gunConfig: GunDataTypes.GunConfig, timeOffset: number): number
+        if gunConfig.DelayPerShot == 0 then
+            return math.round(gunConfig.FireRate*timeOffset)
+        else
+            local t = 0
+            local shots = 0
+            while (t < timeOffset) do
+                for i = 1, gunConfig.BulletsPerShot do
+                    shots += 1
+                    t += ((gunConfig.FireRate*timeOffset)/gunConfig.BulletsPerShot)
+                    if t > timeOffset then
+                        break
+                    end
+                end
+            end
+            return shots
+        end
+    end
+
+    validationTests["FireRateCheck"] = function(player: Player, validationData: validationDataType): boolean
+        local binderInstance = validationData.BinderInstance
+        
+        local gunConfig = validationData.GunConfig
+        local delayPerShot = gunConfig.DelayPerShot
+        local bulletsPerShot = gunConfig.BulletsPerShot
+        local fireRate = gunConfig.FireRate
+
+        if delayPerShot == 0 and fireRate < 3 then
+            local timeOffset = 1/fireRate - marginOfErrorTimeOffset
+            local bulletCount = binderInstance.BulletHistory:GetBulletCountFromTimestampOffset(timeOffset)
+            if bulletCount >= bulletsPerShot/fireRate then
+                return false
+            end
+        else
+            local timeOffset = 0.2
+            local bulletCount = binderInstance.BulletHistory:GetBulletCountFromTimestampOffset(timeOffset)
+            local requiredBulletCount = getBulletCount(validationData.GunConfig, timeOffset)
+            return bulletCount <= requiredBulletCount
+        end
+
+        return true
+    end
+
+    validationTests["RayCastCheck"] = function(player: Player, validationData: validationDataType): boolean
+        local blacklist = BulletHandlerService:GetRaycastBlacklist()
+        local bulletData = validationData.BulletData
+        local direction = bulletData.EndPosition-bulletData.BarrelPosition
+        if direction.Magnitude < 10 then
+            return true
+        end
+        local unitDirection = direction.Unit
+        local rayResult = workspace:Raycast(bulletData.BarrelPosition+unitDirection, direction-unitDirection*3)
+        if rayResult then
+            return false
+        end
+        return true
+    end
+end
```

Comment:
This validation code should almost certainly be it's own module, in fact, any time you feel tempted to use a do end block, there's likely a structure hiding in plain sight.